README.md
# hobart-svg

[![version](https://img.shields.io/pypi/v/hobart-svg?style=flat-square)][pypi]
[![license](https://img.shields.io/pypi/l/hobart-svg?style=flat-square)][pypi]
[![build](https://img.shields.io/circleci/project/github/lace/hobart-svg/master?style=flat-square)][build]
[![docs build](https://img.shields.io/readthedocs/hobart-svg?style=flat-square)][docs build]
[![code style](https://img.shields.io/badge/code%20style-black-black?style=flat-square)][black]

Render polygons, polylines, and mesh cross sections to SVG.

[pypi]: https://pypi.org/project/hobart-svg/
[black]: https://black.readthedocs.io/en/stable/
[build]: https://circleci.com/gh/lace/hobart-svg/tree/master
[docs build]: https://hobart-svg.readthedocs.io/en/latest/


Features
--------

- Render 2D and 3D polygons and polylines to SVG, with automatic computation
  of the bounding rectangle.
- Render cross sections of [lacecore][]-style polygonal meshes.
<!--
- Complete documentation: https://hobart-svg.readthedocs.io/en/stable/
-->

[lacecore]: https://github.com/lace/lacecore


Installation
------------

```sh
pip install hobart-svg
```

Usage
-----

```sh
python -m hobart_svg.cli horizontal-xs \
    --reference vitra_with_xs.dae \
    examples/vitra/vitra_without_materials.obj \
    15 30 45 60
```

```py
from hobart_svg import render_longest_xsection_to_svg
import lacecore
import numpy as np
from polliwog import Plane
import vg

mesh = lacecore.load_obj(filename="mesh.obj", triangulate=True)

plane = Plane(
    reference_point=np.array([0.0, 30.0, 0.0]),
    normal=vg.basis.y
)

render_longest_xsection_to_svg(
    mesh=mesh,
    plane=plane,
    filename="cross_section.svg")
```

```py
from hobart_svg import render_longest_xsection_to_svg
import lacecore
from polliwog import Plane
from tri_again import Scene

mesh = lacecore.load_obj(
    filename="examples/vitra/vitra_without_materials.obj",
    triangulate=True
)
plane = Plane(
    reference_point=np.array([-0.869231, 60.8882, -20.1071]),
    normal=vg.normalize(np.array([0., 0.1, -1.]))
)
xs = render_longest_xsection_to_svg(
    mesh=mesh,
    plane=plane,
    filename="vitra_cross_section.svg"
)

Scene().add_meshes(mesh).add_lines(xs).write("vitra_with_cross_section.dae")
```

## Development

First, [install Poetry][].

After cloning the repo, run `./bootstrap.zsh` to initialize a virtual
environment with the project's dependencies.

Subsequently, run `./dev.py install` to update the dependencies.

[install poetry]: https://python-poetry.org/docs/#installatio


License
-------

The project is licensed under the two-clause BSD license.


CHANGELOG.md
# Changelog

## 1.0.0a0

- Require Python 3.9+
- Update dependencies

## 0.4.1

- Fix issue where dependencies were omitted from 0.4.0.

## 0.4.0

- Bump lacecore
- Note incompatibility with Python 3.9.

## 0.3.0

- Migrate from lace to lacecore.
- Bump polliwog to 1.0.0b10.

## 0.2.2

- Rename package from `hobart` to `hobart-svg`.

## 0.2.1

### Bug fixes

- Update for latest polliwog and lace.
- Declare svgwrite dependency.

### Other maintenance

- Add a test.
- Fix license classifier.


## 0.2.0

Initial release to PyPI.


hobart_svg/cli.py
"""
python -m hobart.cli horizontal-xs \
    --reference vitra_with_xs.dae \
    examples/vitra/vitra_without_materials.obj \
    15 30 45 60
"""

import click


@click.group()
def cli():
    pass


@cli.command()
@click.argument("mesh_path")
@click.argument("heights", nargs=-1, type=float, required=True)
@click.option("-o", "--out", help="Output path")
@click.option(
    "--reference",
    help="When set, write a .dae at this path with the original mesh and the cross section",
)
def horizontal_xs(mesh_path, heights, out, reference):
    """
    Find the horizontal cross section at the given height and write it to an
    SVG file. Optionally write a COLLADA reference with the mesh and cross
    section.
    """
    import os
    import lacecore
    import numpy as np
    from polliwog import Plane
    from tri_again import Scene
    import vg
    from .core import render_longest_xsection_to_svg

    if reference and not reference.endswith(".dae"):
        raise ValueError("reference path should end with .dae")

    mesh = lacecore.load_obj(mesh_path, triangulate=True)

    reference_lines = []

    for height in heights:
        if out is None:
            filename, extension = os.path.splitext(os.path.basename(mesh_path))
            out_path = f"{filename}_cross_section_at_{height}.svg"

        plane = Plane(
            reference_point=np.array([0.0, height, 0.0]), normal=vg.basis.neg_y
        )

        xs = render_longest_xsection_to_svg(
            mesh=mesh, plane=plane, filename=out or out_path
        )

        reference_lines.append(xs)

    if reference:
        Scene().add_meshes(mesh).add_lines(*reference_lines).write(reference)


@cli.command()
@click.argument("measured_body_json_path")
@click.option(
    "--up",
    type=click.Choice(["x", "y", "z", "neg_z", "neg_y", "neg_z"]),
    help="Up vector",
)
@click.option("--flip", is_flag=True, help="Flip the output relative to the plane")
@click.option("-o", "--out", help="Output path")
@click.option("--stroke-color", default="purple", help="Stroke color")
@click.option("--fill-color", default="none", help="Fill color")
def measured_body(measured_body_json_path, out, up, flip, fill_color, stroke_color):
    """
    Find the horizontal cross section at the given height and write it to an
    SVG file. Optionally write a COLLADA reference with the mesh and cross
    section.
    """
    import os
    from missouri import json
    import numpy as np
    from polliwog import Plane, Polyline
    import vg
    from hobart_svg.svg import write_polyline_3d

    measured_body = json.load(measured_body_json_path)

    num_items = len(measured_body["curves"])

    for measurement_name, curve_data in measured_body["curves"].items():
        curve = Polyline.deserialize(
            {"vertices": curve_data["vertices"], "isClosed": curve_data["is_closed"]}
        )

        plane = Plane.fit_from_points(curve.v)

        # Given a pair of these on different meshes, try to produce consistent
        # normals.
        plane = plane.flipped_if(
            not np.array_equal(
                vg.aligned_with(plane.normal, np.array([1, 1, 1])), plane.normal
            )
        )

        if flip:
            plane = plane.flipped()

        if out is None:
            filename, extension = os.path.splitext(
                os.path.basename(measured_body_json_path)
            )
            if num_items == 1:
                out = f"{filename}.svg"
            else:
                out = f"{filename}_{measurement_name}.svg"

        write_polyline_3d(
            polyline=curve,
            filename=out,
            look=plane.normal,
            up=None if up is None else getattr(vg.basis, up),
            polyline_format_attrs={"stroke": stroke_color, "fill": fill_color},
        )


if __name__ == "__main__":
    cli()
